buildSrc/src/main/kotlin/codegen/math/GenerateMath.kt

```diff
@@ -0,0 +1,1027 @@
+		val baseType: ClassName? = null,
+		val componentCount: Int? = null
+	)
+
+
+	private val packageName = "com.kgl.math"
+
+	private val primitiveTypes = listOf(
+		TypeInfo(BOOLEAN),
+		TypeInfo(INT),
+		TypeInfo(LONG),
+		TypeInfo(U_INT),
+		TypeInfo(U_LONG),
+		TypeInfo(FLOAT),
+		TypeInfo(DOUBLE)
+	)
+
+	private val vectorTypes = listOf(
+		TypeInfo(ClassName(packageName, "BooleanVector2"), BOOLEAN, 2),
+		TypeInfo(ClassName(packageName, "BooleanVector3"), BOOLEAN, 3),
+		TypeInfo(ClassName(packageName, "BooleanVector4"), BOOLEAN, 4),
+		TypeInfo(ClassName(packageName, "IntVector2"), INT, 2),
+		TypeInfo(ClassName(packageName, "IntVector3"), INT, 3),
+		TypeInfo(ClassName(packageName, "IntVector4"), INT, 4),
+		TypeInfo(ClassName(packageName, "LongVector2"), LONG, 2),
+		TypeInfo(ClassName(packageName, "LongVector3"), LONG, 3),
+		TypeInfo(ClassName(packageName, "LongVector4"), LONG, 4),
+		TypeInfo(ClassName(packageName, "UIntVector2"), U_INT, 2),
+		TypeInfo(ClassName(packageName, "UIntVector3"), U_INT, 3),
+		TypeInfo(ClassName(packageName, "UIntVector4"), U_INT, 4),
+		TypeInfo(ClassName(packageName, "ULongVector2"), U_LONG, 2),
+		TypeInfo(ClassName(packageName, "ULongVector3"), U_LONG, 3),
+		TypeInfo(ClassName(packageName, "ULongVector4"), U_LONG, 4),
+		TypeInfo(ClassName(packageName, "FloatVector2"), FLOAT, 2),
+		TypeInfo(ClassName(packageName, "FloatVector3"), FLOAT, 3),
+		TypeInfo(ClassName(packageName, "FloatVector4"), FLOAT, 4),
+		TypeInfo(ClassName(packageName, "DoubleVector2"), DOUBLE, 2),
+		TypeInfo(ClassName(packageName, "DoubleVector3"), DOUBLE, 3),
+		TypeInfo(ClassName(packageName, "DoubleVector4"), DOUBLE, 4)
+	)
+
+	private val allComponentNames = listOf("x", "y", "z", "w")
+	private val arithmetics = listOf("plus" to "+", "minus" to "-", "times" to "*", "div" to "/")
+
+
+	@TaskAction
+	fun generate() {
+		commonFunctions()
+		geometricFunctions()
+		vectorTypes()
+		vectorRelationalFunctions()
+	}
+
+	private fun commonFunctions() {
+		(primitiveTypes + vectorTypes).forEach { (type, baseType, componentCount) ->
+			val componentNames = allComponentNames.take(componentCount ?: 0)
+			val mutableType = ClassName(packageName, "Mutable${type.simpleName}")
+
+			val boolVectorType = vectorTypes.find { it.baseType == BOOLEAN && it.componentCount == componentCount }
+
+			val (zero, vectorZero) = literal("0", type, baseType)
+			val (one, vectorOne) = literal("1", type, baseType)
+			val (two, vectorTwo) = literal("2", type, baseType)
+			val (three, vectorThree) = literal("3", type, baseType)
+
+			buildFile(packageName, "${type.simpleName}Common") {
+				indent("\t")
+
+				import("kotlin.math", "abs")
+				import("kotlin.math", "ceil")
+				import("kotlin.math", "floor")
+				import("kotlin.math", "round")
+				import("kotlin.math", "sign")
+				import("kotlin.math", "truncate")
+
+				// abs
+
+				when (baseType) {
+					INT, LONG,
+					FLOAT, DOUBLE -> {
+						extensionFunction(type, "abs") {
```

Comment:
Should this be an extension function? Since stdlib `abs` is not. (There's also an `absoluteValue` extension property.)

Reply:
Not making it an extension calls the use of `absAssign` into question.

Reply:
Ah. Then everything else concerning `opAssign` too.
I guess we can leave this can of worms closed.

Reply:
I've revisited this since breaking the pattern set by the stdlib was bothering me, and stdlib top-level functions that take a parameter have been changed to `op(n: immutable): immutable` / `opInPlace(n: mutable)` rather than being extension functions.